linera-witty/src/type_traits/implementations/std/tuples.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementations of the custom traits for the tuple types.
+
+use crate::WitType;
+use frunk::HList;
+
+macro_rules! impl_wit_traits {
+    ($head_name:ident : $head_type:ident, $( $tail_names:ident : $tail_types:ident ),*) => {
+        impl_wit_traits!($head_name: $head_type | $( $tail_names: $tail_types ),*);
+    };
+
+    ($( $names:ident : $types:ident ),* |) => {
+        impl_wit_traits!(@generate $( $names: $types, )*);
+    };
+
+    (
+        $( $names_to_generate:ident : $types_to_generate:ident ),* |
+        $next_name:ident : $next_type:ident $( , $queued_names:ident : $queued_types:ident )*
+    ) => {
+        impl_wit_traits!(@generate $( $names_to_generate: $types_to_generate, )*);
+        impl_wit_traits!(
+            $( $names_to_generate: $types_to_generate, )*
+            $next_name: $next_type | $( $queued_names: $queued_types ),*);
+    };
+
+    (@generate $( $names:ident : $types:ident, )*) => {
+        impl<$( $types ),*> WitType for ($( $types, )*)
+        where
+            $( $types: WitType, )*
+            HList![$( $types ),*]: WitType,
+        {
+            const SIZE: u32 = <HList![$( $types ),*] as WitType>::SIZE;
+
+            type Layout = <HList![$( $types ),*] as WitType>::Layout;
+        }
+    };
+}
+
+impl_wit_traits!(
```

Comment:
It would work, it's the same thing. I think it's just a matter of style, I prefer it this way because the repetition pattern will also be used with other macros, and I believe is less prone to errors.

But I can change it if you want.

Reply:
Ok let's keep the general idea but can you define 3 separate macros instead of using magic tokens `|` and `@generate` to differentiate between 3 kinds of calls? (And if not, at least make `impl_wit_traits!(@generate ...)` a separate one.)